### PR TITLE
[DNM] Add scatter/gather IO support to librbd API

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

File: src/librbd/librbd.cc [hunk 1]
```diff
@@ -61,6 +62,16 @@ namespace {
 
 TracepointProvider::Traits tracepoint_traits("librbd_tp.so", "rbd_tracing");
 
+buffer::raw* create_write_raw(librbd::ImageCtx *ictx, const char *buf,
+                              size_t len) {
+  // cannot allow the cache (if enabled) to retain reference to user memory
+  if (ictx->cache) {
+    return buffer::copy(buf, len);
+  } else {
+    return buffer::create_static(len, const_cast<char*>(buf));
```
Comment (dillaman) [inline src/librbd/librbd.cc]: Messenger code could still reference memory after completing op(?)

Comment (yuyuyu101) [inline src/librbd/librbd.cc]: no, messenger also use bufferlist to trace memory, but this memory is owner by caller. So if caller free this memory, messenger also have no idea to reference

Comment (dillaman) [inline src/librbd/librbd.cc]: indeed -- it was Sage's example of a PG mapping update and the messenger threads racing between old and new PGs. In a perfect world, I would expect that once librados ACKs, the memory wouldn't be used again. Baring that, the C API just a needs to always copy at least once.

Comment (yuyuyu101) [inline src/librbd/librbd.cc]: Yes, I hope we can support cancel_message to do this... Currently messenger exposes too little

**Comment by dillaman** [comment]
Will hold to post-Jewel

**Comment by dillaman** [comment]
Refactored version of this is included in the forthcoming refactoring of the IO path PR

**Comment by dillaman** [comment]
This will be addressed with a different PR that includes an IO pathway refactoring.

**Comment by jinweiyi** [comment]
@dillaman, I  am working on the librbd IO PR mainly base on your commits within this pull request.

**Comment by dillaman** [comment]
@jinweiyi FYI - the scatter/gather IO work is essentially already complete and will be included with the Luminous release. There is a possibility that this patchset will be re-opened for backport to Kraken/Jewel in support of the TCMU+librbd iSCSI implementation.

**Pull request closed without merging**